crates/sui-core/src/overload_monitor.rs

```diff
@@ -0,0 +1,292 @@
+}
+
+impl AuthorityOverloadInfo {
+    pub fn set_overload(&self, load_shedding_percentage: u32) {
+        self.is_overload.store(true, Ordering::Relaxed);
+        self.load_shedding_percentage
+            .store(min(load_shedding_percentage, 100), Ordering::Relaxed);
+    }
+
+    pub fn clear_overload(&self) {
+        self.is_overload.store(false, Ordering::Relaxed);
+        self.load_shedding_percentage.store(0, Ordering::Relaxed);
+    }
+}
+
+// Monitors the overload signals in `authority_state` periodically, and updates its `overload_info`
+// when the signals indicates overload.
+pub async fn overload_monitor(
+    authority_state: Weak<AuthorityState>,
+    config: OverloadThresholdConfig,
+) {
+    info!("Starting system overload monitor.");
+
+    loop {
+        let authority_exist = check_authority_overload(&authority_state, &config);
+        if !authority_exist {
+            // `authority_state` doesn't exist anymore. Quit overload monitor.
+            break;
+        }
+        sleep(config.overload_monitor_interval).await;
+    }
+
+    info!("Shut down system overload monitor.");
+}
+
+// Checks authority overload signals, and updates authority's `overload_info`.
+// Returns whether the authority state exists.
+fn check_authority_overload(
+    authority_state: &Weak<AuthorityState>,
+    config: &OverloadThresholdConfig,
+) -> bool {
+    let authority_arc = authority_state.upgrade();
+    if authority_arc.is_none() {
+        // `authority_state` doesn't exist anymore.
+        return false;
+    }
+
+    let authority = authority_arc.unwrap();
+    let queueing_latency = authority
+        .metrics
+        .execution_queueing_latency
+        .latency()
+        .unwrap_or_default();
+    let txn_ready_rate = authority.metrics.txn_ready_rate_tracker.lock().rate();
+    let execution_rate = authority.metrics.execution_rate_tracker.lock().rate();
+
+    let (is_overload, load_shedding_percentage) =
+        check_overload_signals(config, queueing_latency, txn_ready_rate, execution_rate);
+    if is_overload {
+        authority
+            .overload_info
+            .set_overload(load_shedding_percentage);
+    } else {
+        authority.overload_info.clear_overload();
+    }
+
+    authority
+        .metrics
+        .authority_overload_status
+        .set(is_overload as i64);
+    authority
+        .metrics
+        .authority_load_shedding_percentage
+        .set(load_shedding_percentage as i64);
+    true
+}
+
+// Calculates the percentage of transactions to drop in order to reduce execution queue.
+// Returns the integer percentage between 0 and 100.
+fn calculate_load_shedding_percentage(txn_ready_rate: f64, execution_rate: f64) -> u32 {
+    // When transaction ready rate is practically 0, we aren't adding more load to the
+    // execution driver, so no shedding.
+    // TODO: consensus handler or transaction manager can also be overloaded.
+    if txn_ready_rate < 1e-10 {
```

Comment:
This seems like an oddly low threshold. I would not be surprised to see a lot of numerical instability at rates well above 1e-10, (say, 1e-5). So it's not obvious to me that we wouldn't see the system thrashing in and out of overload protection at very low traffic rates.

It would seem that a much higher threshold, say 10 or even 100 tps should be safe?

Reply:
ok, makes sense for now